Support Python 3

bertsky opened the pull request:
This fixes #6 and adds support for Python 2 __and__ 3 by (doing some preparation and parameterization for) `futurize`.

I wish we had unit tests here (like `make check` in leptonica itself) – so I don't know whether this breaks anything. I _do_ get some of the warnings mentioned in #7, but I believe this is because my local `liblept5` or `libleptonica-dev` does not match the generated code here.

Along with this PR, maybe we could clarify what the release policy for PyPI is supposed to be. I can see `0.8` via both `pip search pyleptonica` and `pip3 search pyleptonica`, but cannot `pip[3] install`:
```
ERROR: No matching distribution found for pyleptonica
```

Commit:
replace os.stderr with sys.stder (and import sys)

Commit:
fix invalid del

Commit:
use proper PEP-263 encoding comments

Commit:
futurize -nw0 -x lib2to3.fixes.fix_dict leptonica

kba commented:
Up-to-date python bindings for leptonica would offer a whole range of image processing options for @OCR-D Just porting some of leptonica test/example scripts to Python can help in various tasks like segmentation, dewarping etc.

@jsbueno If we can help with maintenance, deployment, testing, etc. let me know.

bertsky commented:
This does not work yet:

```
import leptonica
leptonica.functions.pixRead('image.pnm')
Error in findFileFormatStream: failed to read first 12 bytes of file
Error in pixReadStream: Unknown format: no pix returned
Error in pixRead: pix not read
Leptonica PIX object
Not initiated or destroyed
```
(same if I use `leptonica.readfile.pixRead`)

After that, I always get an extra:
```
Traceback (most recent call last):
  File "leptonica/leptonica_structures.py", line 80, in __del__
    if self._needs_del and hasattr(functions, cls.__name__.lower() + "Destroy"):
  File "leptonica/leptonica_structures.py", line 46, in __getattribute__
    raise ValueError("Object no longer exists")
ValueError: Object no longer exists
```

And the provided extra PIL interface is not Python 3-ready yet:
```
leptonica.PILImageToPix(PIL.Image.open('image.pnm'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "leptonica/leptonica_utils.py", line 62, in PILImageToPix
    img_data = pil_img.tostring()
  File "PIL/Image.py", line 722, in tostring
    "tostring() has been removed. Please call tobytes() instead."
NotImplementedError: tostring() has been removed. Please call tobytes() instead.
```

I have installed `libleptonica 1.75.3` (from Feb 2018), whereas 6cc0af60b2302ca0b99ecdf54ca282e21e3d546d is from 2015 (which I find strange, because it already references header files that were introduced much later, like `ptra.h watershed.h regutils.h` from 2016). There is no versioning here (and no mention of versions in the changelog). Hard to guess what version of libleptonica this currently expects.

So I gathered what must be done next is parsing (with Python 3) an up-to-date version of the leptonica header and source files, as described in the `README`:

> I versioned the generated files as well so that most people just have
to copy the .py files around. If you want to hack the code, or generate 
wrappers for other versions of Leptonica, you have to execute
the `leptonica_header_parser.py`, `leptonica_function_parser.py` and
`leptonica_enum_parser.py` as stand alone programs, after tunning a
config.py from the `config_template.py` file.

So I built leptonica from git (version tag is `1.80.0` but it has not been released yet), and ran the header parser against that (`python leptonica_header_parser.py`).

The first obstacle is another Python2-ism:

```
python leptonica_header_parser.py 
Traceback (most recent call last):
  File "leptonica_header_parser.py", line 417, in <module>
    main(all_headers)
  File "leptonica_header_parser.py", line 406, in main
    class_list = order_classes(structs)
  File "leptonica_header_parser.py", line 374, in order_classes
    for struct in structs.keys()[:]:
TypeError: 'dict_keys' object is not subscriptable
```

But this can easily be replaced by `list()` around the generator. 

Next we get the following:

```
{'BMP_FH', 'BMP_IH'}

{'BOX', 'L_QUEUE', 'JBCLASSER', 'L_DNAA', 'FPIX', 'NUMA', 'L_DNAHASH', 'PIXAA', 'L_DNA', 'L_PTRA', 'L_PTRAA', 'PIXACC', 'BMP_INFOHEADER', 'CCBORD', 'JBDATA', 'DOUBLELINKEDLIST', 'L_BYTEBUFFER', 'L_STACK', 'BOXA', 'FPIXA', 'RGBA_QUAD', 'GPLOT', 'SEL', 'SELA', 'L_KERNEL', 'PIX', 'PIXA', 'NUMAA', 'L_BYTEA', 'L_BMF', 'PIXACOMP', 'PTAA', 'CCBORDA', 'BOXAA', 'L_HEAP', 'L_WSHED', 'PTA', 'L_REGPARAMS', 'L_SUDOKU', 'PIXCOLORMAP', 'PIXCOMP', 'PIXTILING', 'DPIX', 'BMP_FILEHEADER', 'SARRAY'}
Traceback (most recent call last):
  File "leptonica_header_parser.py", line 417, in <module>
    main(all_headers)
  File "leptonica_header_parser.py", line 406, in main
    class_list = order_classes(structs)
  File "leptonica_header_parser.py", line 386, in order_classes
    "Could not get Struct classes in order")
Exception: Could not get Struct classes in order
```

This looks like a discrepancy between the explicit list of header files and the actual include tree in the sources (which have of course changed since 2015):

```python
all_headers = ['array.h', 'bbuffer.h', 'gplot.h', 'pix.h', 
'regutils.h', 'bmf.h', 'heap.h', 'ptra.h', 'stack.h', 'bmp.h',
'list.h', 'queue.h', 'sudoku.h', 'array.h', 'ccbord.h', 'jbclass.h',
'morph.h', 'watershed.h']
```

For example, I miss `'rbtree.h', 'bmfdata.h', 'bilateral.h', 'imageio.h', 'recog.h', 'stringcode.h', 'dewarp.h'` in that list.

But I must confess I don't grasp how `order_classes()` is supposed to work at all.

Can anyone help with insights?

Commit:
fix and update header parser...

- update/complete list of header files
- add 64-bit integer types
- also add l_ok (most common return type)
- traverse union besides struct
- update list of exceptional typedef names
  (not identical to uppercase struct)
- pass list of exceptional names into function
  ordering class dependencies as well
- avoid generator slicing and destructive iterator access
- avoid macro-computed names
- no exception for L_TIMER (but add to native types)
- handle mutual/circular dependencies as well:
  - while ordering, break circles
  - while generating, first aggregate
    all ctypes and Python class declarations,
    then aggregate their member definitions
    to prevent refering to yet undefined types
- LeptonObject: avoid importing leptonica.functions
  during del (which might be too late) looking
  for destructors;
  instead, find the destructor function in the
  constructor and store it as member
- aggregate all defined Python types (and their
  abbreviated forms) into __all__ for export

Commit:
fix and update functions parser...

- update/complete list of source files
  (deleted/renamed/added) since 1.67
- handle/ignore copyright comments
- adapt to file comments in new style/syntax:
  - find and keep both file-wide doxygen comments,
    but remove unformatted section header comments
  - find function doxygen comments more reliably
  - when converting doxygen to docstring,
    remove \brief and \file, but keep the
    \param declaration using colons
  - avoid running into structs etc
- functions which return a ctypes object pointer
  must convert null to None before instantiating
  a Python object
- convert char* args to c_char_p instead of c_void_p
- convert str instances to (UTF-8) bytes for c_char_p
- avoid exporting memory manager related functions

Commit:
fix and update enum parser...

- allow nun-numeric (alias) values
- update/complete list of source files
- fix/adapt order of title vs doc comments
  (new comment style)
- fix/improve stripping comments
- fix adding enums to __all__

Commit:
update from leptonica 1.80.0 (git 948e52d700)

Commit:
PILImageToPix: bytes instead of str

bertsky commented:
:relieved::sweat: Finally got it working!

> Error in findFileFormatStream: failed to read first 12 bytes of file
> Error in pixReadStream: Unknown format: no pix returned
> Error in pixRead: pix not read

The cause for this was the change from strings to bytes between Python 2 and 3, so the `ctypes` wrapper would only pass the first byte of any string param.

> Not initiated or destroyed
> ValueError: Object no longer exists

The bug here was missing NULL-None conversion.

But there were much bigger problems under the sheets: 
- leptonica sources have changed their comment styles/conventions
- leptonica sources have been re-structured (spliting/renaming/adding/removing files)
- more/other basic types, missing unions besides structs
- handling mutual/circular dependencies between structs and their members

Generally though I must say at this point that the whole approach here is a PITA. It's just wrong (and nearly impossible to maintain) to:
1. parse C code from Python (with ad-hoc self-made "parsers"!) to extract API and docstrings
2. maintain lists of header files (once for structs and once for enums) and source files to process
3. have no versioning or ABI synchronization mechanism whatsoever

If anyone knows how to do this _programmatically_ (perhaps using Cython, and for the docstrings I guess there must be some pipeline directly based on doxygen) – please help!

bertsky commented:
> But there were much bigger problems under the sheets:
> 
> * leptonica sources have changed their comment styles/conventions
> * leptonica sources have been re-structured (spliting/renaming/adding/removing files)
> * more/other basic types, missing unions besides structs
> * handling mutual/circular dependencies between structs and their members
> 
> Generally though I must say at this point that the whole approach here is a PITA. It's just wrong (and nearly impossible to maintain) to:
> 
> 1. parse C code from Python (with ad-hoc self-made "parsers"!) to extract API and docstrings
> 2. maintain lists of header files (once for structs and once for enums) and source files to process
> 3. have no versioning or ABI synchronization mechanism whatsoever
> 
> If anyone knows how to do this _programmatically_ (perhaps using Cython, and for the docstrings I guess there must be some pipeline directly based on doxygen) – please help!

So what do we do now, @kba? This library is not maintained anymore, and I can see no replacement. The problems detailed above are severe, but we could at least start working on them, if we switch to our own fork. Therefore, I suggest:

- using OCR-D/pyleptonica on GH and `ocrd-fork-pyleptonica` on PyPI as base fork maintained by OCR-D
- merging this PR, perhaps with some more changes to README (Markdown, update, typos) and setup.py (link, dist name, version)
- repeating the parsing step once more for the newest Leptonica 1.83
- starting a replacement mechanism for the manual parser and ctypes code; perhaps something based on [SIP](https://realpython.com/python-bindings-overview/#sip) or [CFFI](https://cffi.readthedocs.io/en/latest/overview.html)?

bertsky commented:
I also have an idea for CI regression tests: In Leptonica itself, there are a couple of these (under `prog/*_reg.c`), which must be run with `generate` (to create the reference output files) and `compare` (to create new output and validate against the reference). There is also an entry point `alltests_reg` for the whole suite. So what we could do for our Python wrapper is the following:
- port each `leptonica/prog/*_reg.c` into some `pyletonica/test/*_reg.py`
- run `generate` on leptonica and `compare` on pyleptonica tests

(Perhaps we could also add the reference output into the repo right away, and only update each time the C code gets re-parsed...)

kba commented:
> So what do we do now, @kba? This library is not maintained anymore, and I can see no replacement. The problems detailed above are severe, but we could at least start working on them, if we switch to our own fork. Therefore, I suggest:

Sure, we can host the fork and do release management in OCR-D, but I am unsure whether I can be of much help with the actual bindings, your code review does not exactly inspire confidence. How does tesserocr wrap the C code, could that be an inspiration for a rewrite, since they work very well in my experience? I am all for leptonica python bindings but want to make sure we're not biting off more than we can chew.

bertsky commented:
> your code review does not exactly inspire confidence

For the moment it could work (I guess – we only know once we test), but going forward, the effort to keep track of changes in Leptonica is high. And we need at least some versioning.

> How does tesserocr wrap the C code, could that be an inspiration for a rewrite, since they work very well in my experience?

It manually declares and re-exports (+ wraps, if necessary, like Pix→PIL) all relevant functions and data structures via Cython. Pyleptonica currently does the same, but with ctypes. And while tesserocr has its own docstrings (at some point derived from the C comments in Tesseract), pyleptonica retrieves them automatically (if in a fragile way). Using CFFI or SIP would alleviate all that, it seems. But it _is_ a huge bite to set it up properly.
